modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/IgniteRebalanceIteratorImpl.java

```diff
@@ -96,6 +99,11 @@ private synchronized void advance() throws IgniteCheckedException {
         return historicalIterator != null && historicalIterator.contains(partId);
     }
 
+    /** {@inheritDoc} */
+    @Override public synchronized boolean allHistoricalPartitionsDone() {
+        return historicalIterator == null || reachedHistEnd;
```

Comment:
Why we need reachedHistEnd ?
Can it be replaced with historicalIterator  != null && !historicalIterator.hasNext() ?

Reply:
Unfortunately, no. It cannot be replaced. The reason is `WALHistoricalIterator#advance()`. Being more precise, this method at the first step nullifies the field `next`. If this method throws an exception while it iterates data entries, the field `next` still equals to 'null', and so `hasNext()` returns `false`. It looks weird to me. Perhaps we need to rework `WALHistoricalIterator#advance()`. What do you think?